cpp/src/neighbors/detail/tiered_index.cuh

```diff
@@ -0,0 +1,396 @@
+      storage->dataset.data() + ann_rows() * storage->dim, bfknn_rows(), storage->dim);
+
+    brute_force::index<value_type> bfknn_index(
+      res, bfknn_dataset_view, norms_view, build_params.metric);
+
+    // if we don't have an ANN index, just return the bfknn results right away
+    if (!ann_index) {
+      brute_force::search(res,
+                          brute_force::search_params(),
+                          bfknn_index,
+                          queries,
+                          neighbors,
+                          distances,
+                          sample_filter);
+      return;
+    }
+
+    // query the ann/bfknn results into temporary memory. Since the knn_merge_parts
+    // api expects inputs to be in a contiguous layout - create a single temporary
+    // and then slice from it for the ann/bfknn search
+    auto n_queries      = distances.extent(0);
+    auto k              = distances.extent(1);
+    auto temp_distances = raft::make_device_matrix<value_type, int64_t>(res, 2 * n_queries, k);
+    auto temp_neighbors = raft::make_device_matrix<int64_t, int64_t>(res, 2 * n_queries, k);
+
+    // search the ann index
+    search_fn(
+      res,
+      search_params,
+      *ann_index,
+      queries,
+      raft::make_device_matrix_view<int64_t, int64_t>(temp_neighbors.data_handle(), n_queries, k),
+      raft::make_device_matrix_view<value_type, int64_t>(
+        temp_distances.data_handle(), n_queries, k),
+      sample_filter);
+
+    // search the bfknn index
+    auto offset          = n_queries * k;
+    auto bfknn_neighbors = raft::make_device_matrix_view<int64_t, int64_t>(
+      temp_neighbors.data_handle() + offset, n_queries, k);
+    auto bfknn_distances = raft::make_device_matrix_view<value_type, int64_t>(
+      temp_distances.data_handle() + offset, n_queries, k);
+    brute_force::search(res,
+                        brute_force::search_params(),
+                        bfknn_index,
+                        queries,
+                        bfknn_neighbors,
+                        bfknn_distances,
+                        sample_filter);
+
+    if (!distance::is_min_close(build_params.metric)) {
+      // knn_merge_parts doesn't currently support InnerProduct distances etc
+      // instead negate here and then undo after
+      raft::linalg::map(res,
+                        temp_distances.view(),
+                        raft::mul_const_op<value_type>(-1),
+                        raft::make_const_mdspan(temp_distances.view()));
+    }
+
+    // merge results from ann_index/bfknn together, translating the bfknn ids
+    auto stream                  = raft::resource::get_cuda_stream(res);
+    int64_t host_translations[2] = {0, static_cast<int64_t>(ann_rows())};
+    auto device_translations     = raft::make_device_vector<int64_t>(res, 2);
+    raft::copy(device_translations.data_handle(), host_translations, 2, stream);
+
+    cuvs::neighbors::detail::knn_merge_parts(temp_distances.data_handle(),
```

Comment:
Not sure how heavy this kernel is, but I can imagine it's instantiated several times. Can opportunity to reuse instantiations here?

Reply:
this is definitely instantiated many times: in addition to here, `knn_merge_parts` is used  in the snmg code, bfknn and the sparse bfknn.

I'll add an instantiation here -